coverage/docs/design.md

```diff
@@ -0,0 +1,113 @@
+
+Prow has handlers for different github events. We add a pre-submit prow job that is triggered by any new commit to a PR to run test coverage on the new build to compare it with the master branch and previous commit for pre-submit coverage. We add a post-submit prow job that is triggered by merge events, to run test coverage on the nodes of master branch. Test coverage data on the master branch is supplied to TestGrid for displaying the coverage change over time, as well as serve as the basis of comparison for pre-submit coverage mentioned in the pre-submit senario.
+
+Here is the step-by-step description of the pre-submit and post-submit workflows
+
+##Pre-submit workflow
+Runs code coverage tool to report coverage change in a new commit or new PR
+1. Developer submit new commit to an open PR on github
+2. Matching pre-submit prow job is started 
+3. Test coverage profile generated
+4. Calculate coverage changes. Compare the coverage file generated in this cycle against the most recent successful post-submit build. Coverage file for post-submit commits were generated in post-submit workflow
+5. Use PR data from github, git-attributes, as well as coverage change data calculated above, to produce a list of files that we care about in the line-by-line coverage report. produce line by line coverage html and add link to covbot report. Note that covbot is the robot github account used to report code coverage change results.
+6. Let covbot post presubmit coverage on github, under that conversation of the PR. 
+7. When coverage threshold is enforced, block PR from merging by making this prow job 'required' and return with a code other than 0
+
+##Post-submit workflow
+Produces & stores coverage profile for later presubmit jobs to compare against
+1. A PR is merged
+2. Post-submit prow job started
+3. Test coverage profile generated. Completion marker generated upon successful run. Both stored as prow artifacts.
+    - Completion marker is used by later pre-submit job when searching for a healthy and complete code coverage profile in the post-submit jobs
+
+##Periodical workflow
+Produces periodical coverage result as input for TestGrid
+1. Periodical prow job starts periodically based on the specification in prow job config
+2. Generate coverage profile
+3. Generate / store per-file coverage data
+  - Stores in the XML format, that is used by TestGrid, and dump it in prow artifacts directory 
+  
+##Prow Configuration File
+As mentioned earlier, we use configuration file to store repository specific information. Below is an example that contains the args that will be supplied to the coverage container
+```$xslt
+  - name: pull-knative-serving-go-coverage
+    labels:
+      preset-service-account: "true"
+    agent: kubernetes
+    context: pull-knative-serving-go-coverage
+    always_run: true
+    rerun_command: "/test pull-knative-serving-go-coverage"
+    trigger: "(?m)^/test (all|pull-knative-serving-go-coverage),?(\\s+|$)"
```

Comment:
You can delete the agent, context, rerun_command, and trigger fields now.

Reply:
Done.